src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java

```diff
@@ -0,0 +1,203 @@
+ * limitations under the License.
+ */
+
+package com.hivemq.embedded.internal;
+
+import com.codahale.metrics.MetricFilter;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Injector;
+import com.hivemq.HiveMQServer;
+import com.hivemq.bootstrap.ioc.GuiceBootstrap;
+import com.hivemq.common.shutdown.HiveMQShutdownHook;
+import com.hivemq.common.shutdown.ShutdownHooks;
+import com.hivemq.configuration.ConfigurationBootstrap;
+import com.hivemq.configuration.HivemqId;
+import com.hivemq.configuration.info.SystemInformationImpl;
+import com.hivemq.configuration.service.FullConfigurationService;
+import com.hivemq.embedded.EmbeddedHiveMQ;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.extension.sdk.api.annotations.Nullable;
+import com.hivemq.persistence.PersistenceStartup;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+/**
+ * @author Georg Held
+ */
+public class EmbeddedHiveMQImpl implements EmbeddedHiveMQ {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(EmbeddedHiveMQImpl.class);
+
+    private final @NotNull SystemInformationImpl systemInformation;
+    private final @NotNull MetricRegistry metricRegistry;
+    private @Nullable FullConfigurationService configurationService;
+    private @Nullable Injector injector;
+
+    private volatile @NotNull State currentState;
+    private volatile @Nullable CompletableFuture<Void> startFuture;
+    private volatile @Nullable CompletableFuture<Void> stopFuture;
+
+    EmbeddedHiveMQImpl(
+            final @Nullable File conf, final @Nullable File data, final @Nullable File extensions) {
+
+        currentState = State.NOT_STARTED;
+        systemInformation = new SystemInformationImpl(true, true, conf, data, extensions);
+        // we create the metric registry here to make it accessible before start
+        metricRegistry = new MetricRegistry();
+    }
+
+    private enum State {
+
+        NOT_STARTED,
+        STARTING,
+        RUNNING,
+        STOPPING,
+        STOPPED;
+
+        private boolean isHigher(final @NotNull State other) {
+            return this.ordinal() > other.ordinal();
+        }
+
+        private boolean isLower(final @NotNull State other) {
+            return this.ordinal() < other.ordinal();
+        }
+    }
+
+    private synchronized void advanceState(final @NotNull State newState) {
+        if ((currentState == State.STOPPED && newState == State.NOT_STARTED)
+                || newState.isHigher(currentState)) {
+            log.debug("Advancing EmbeddedHiveMQState from \"{}\" to \"{}\"", currentState, newState);
+            currentState = newState;
+        }
+    }
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> start() {
+        advanceState(State.NOT_STARTED);
```

Comment:
why is this necessary? the sate is already NOT_STARTED, if it is not -> else

Reply:
For the restart case